### DISCUSSION TITLE
Source __version__ defined inside __init__.py

### DISCUSSION THREAD

**Discussion by bede** [pyinstaller/pyinstaller#7321] [Help] [answered]
Greetings!
I have a package following the common `src` layout that is now working great with PyInstaller:
```
src/pkgname/cli.py
```
 However unlike the pip installed version, the PyInstaller version can't see `__version__` which I define inside:
```
src/pkgname/__init__.py
```
How do I tell PyInstaller to run `__init__.py` and thus set the `__version__` so it's available to my PyInstalled CLI app ? I already tried adding it to the first argument of `Analysis()` unsuccessfully

My spec file looks like this:

```python
# -*- mode: python ; coding: utf-8 -*-
from PyInstaller.utils.hooks import collect_submodules

hiddenimports = []
hiddenimports += collect_submodules('pockets')


block_cipher = None


a = Analysis(
    ['src/pkgname/cli.py'],
    pathex=[],
    binaries=[],
    datas=[('src/pkgname/data', 'data')],
    hiddenimports=hiddenimports,
    hookspath=[],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)

to_rem = ["lib.cpython-310-darwin.so", "lib.cpython-310-x86_64-linux-gnu.so"]

for val in to_rem:
    for b in a.binaries:
          nb = b[0]
          if str(nb).endswith(val):
                print("removed  " + b[0])
                a.binaries.remove(b)

pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)

exe = EXE(
    pyz,
    a.scripts,
    a.binaries,
    a.zipfiles,
    a.datas,
    [],
    name='pkgname',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    upx_exclude=[],
    runtime_tmpdir=None,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
)
```

Thanks!

**Comment by rokm** [comment, marked as answer]
Your `pkgname/__init__.py` is ran when your entry-point script (`pkgname/cli.py`) imports `pkgname`. So you'll have to provide a bit more details on what problem you are facing; in particular, how do you define `__version__` in your `__init__.py`, and how are you trying to access it in your (PyInstaller-frozen) `cli.py`? Perhaps if you can provide minimum examples of `__init__.py` and `cli.py`?

**Reply by bede** [reply]
Thank you @rokm !

Very useful to hear that `__init__.py` is being executed. I'm using a CLI library called [defopt](https://github.com/anntzer/defopt) which [uses importlib to get the module version](https://github.com/anntzer/defopt/blob/bffadb43ad266d172b863a1fdc87b4e9f9096330/lib/defopt.py#L350) and automatically make a available a `--version` CLI option which prints the package version. This behaviour is working with CPython via pip install but not with PyInstaller, and I assumed this is because `__version__` is not being set – I could be mistaken. In light of this and the below excerpts, have you any ideas what might be the issue? The repository is [here](https://github.com/globalPathogenAnalysisService/gpas-cli).

[`__init__.py`](https://github.com/GlobalPathogenAnalysisService/gpas-cli/blob/main/src/gpas/__init__.py)
```python
import importlib.metadata
from pathlib import Path

__version__ = importlib.metadata.version("gpas")
pkg_dir = Path(__file__)
data_dir = pkg_dir.parent / "data"
```
[cli.py](https://github.com/GlobalPathogenAnalysisService/gpas-cli/blob/main/src/gpas/cli.py):
```python
import defopt

…

def main():
    defopt.run(
        {
            "validate": validate_wrapper,
            "upload": upload_wrapper,
            "status": status,
            "download": download,
        },
        no_negated_flags=True,
        strict_kwonly=False,
        short={},
    )


if __name__ == "__main__":
    multiprocessing.freeze_support()
    main()

```

**Reply by bede** [reply]
Helped by your message I've got it working by adding

```python
from gpas import __version__
```

Which defopt infers automatically for the package. Problem solved! Thanks @rokm : )

**Reply by rokm** [reply]
Instead of explicitly adding `from gpas import __version__` to the code, you could also try adding `gpas` to the `hiddenimports` in your spec file; the problem was likely that `gpas` is not directly imported anyhwere in the code.

**Reply by bede** [reply]
This is a great suggestion but didn't work for me unfortunately. Agree it would be nice not to modify the source to get this working.

**Reply by rokm** [reply]
Sorry, I misread your solution. 

So you replaced `__version__ = importlib.metadata.version("gpas")` with `from gpas import __version__`, right? In that case, it was not a missing/hidden import, but likely missing metadata for `gpas` that was the problem.

You can collect metadata using the [`copy_metadata` hook utility function](https://pyinstaller.org/en/stable/hooks.html#PyInstaller.utils.hooks.copy_metadata) in your spec file:

``` spec
from PyInstaller.utils.hooks import copy_metadata

...

a = Analysis(
    ...
    datas=[('src/pkgname/data', 'data')] + copy_metadata('gpas'),
    ...
```

EDIT: although that does require metadata to be present (e.g., `pip install`'ed package). It probably will not work from within source tree.

**Reply by rokm** [reply]
Eh, nevermind...

I took a closer look at `defopt` (the `_get_version1` function that you linked earlier), and it probably fails because when you freeze your `gpas/cli.py`, it effectively becomes a stand-alone piece of python code (instead of submodule of gpas package, as it is originally). So it hits the `module_name == '__main__'` codepath, and then it probably does not have `__spec__` available, so it goes to `return f_globals.get('__version__')` clause. 

That's why adding `from gpas import __version__` to `cli.py` helps, as it puts `__version__` into globals.

I suppose to make things work as originally attempted, we would need to implement equivalent of `python -m gpas.cli` mode as opposed to treating the given .py file as an entry-point script (which basically decouples it from the parent package).

**Reply by bede** [reply]
Thanks, I really appreciate your help with this. I spent ages trying to get PyInstaller working at all with defopt—which I use for all my CLIs—and this was the last small mystery (which I've so far encountered).
